Add Gotify component

benjmarshall opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new platform for the notify component. [Gotify](https://gotify.net/) is an open source, self-hosted push notification server.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/18520

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

benjmarshall committed:
Add Gotify component. Update CODEOWNERS and .coveragerc for new component.

homeassistant commented:
Hi @benjmarshall, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#53050).

Once you do that we will be able to review and accept this pull request.

Thanks!

milanmeu commented on homeassistant/components/gotify/notify.py:
```diff
@@ -0,0 +1,74 @@
+
+        request_json = json.dumps(request_dict)
+
+        resp = requests.request(
+            method="POST", url=request_url, headers=headers, data=request_json
+        )
```
All communication to external devices or services must be wrapped in an external Python library hosted on PyPI.
https://developers.home-assistant.io/docs/creating_platform_code_review#5-communication-with-devicesservices

benjmarshall replied:
Understood. A quick search shows an existing library on pypi which should be suitable so i will look at translating the remote connection to that library.

milanmeu commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+from .const import DOMAIN
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {vol.Required(CONF_TOKEN): cv.string, vol.Required(CONF_URL): cv.string}
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
```
New integrations that communicate with devices and/or services should be configurable with a Config Flow instead of YAML.

For more detail read [ADR-0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision)

benjmarshall replied:
Ok, I will look at how to set this up.

benjmarshall replied:
Hi, I've attempted to migrate this new component over to a Config Flow. This is new to me so I have pushed these changes onto a new branch and would like someone to informally review and provide and guidance/suggestions if things need improved before I merge those changes onto this branch for the PR. The new component has been tested locally and functionally works, the config flow is tested and also seems to work fine. @milanmeu could you please take a quick look at [these changes](https://github.com/benjmarshall/core/commit/4532ed1450db94c2506ac52e66034283239deddb) and provide any advice you can or let me know if you think it's ready to merge into this PR?

TheZoker replied:
I tested this locally as well and it works for me 👍

milanmeu replied:
> @milanmeu could you please take a quick look at [these changes](https://github.com/benjmarshall/core/commit/4532ed1450db94c2506ac52e66034283239deddb) and provide any advice you can or let me know if you think it's ready to merge into this PR?

Here is my comment: https://github.com/benjmarshall/core/pull/5.

Merge `gotify_config_flow` into `add_gotify_integration` so other users/reviewers can see what's happening.

benjmarshall commented on homeassistant/components/gotify/manifest.json:
```diff
@@ -0,0 +1,7 @@
+  "name": "Gotify",
+  "documentation": "https://www.home-assistant.io/integrations/notify_events",
+  "codeowners": ["@benjmarshall"],
+  "iot_class": "cloud_push"
```
Gotify is self-hosted but doesn't have to be on the local network, therefore I think it is most correctly called a "cloud" service. I looked at the class of other similar integrations such as the pushbullet and pushover notification services and the matrix chat integration, the later of which could also be self hosted and can be used as a notification service, and they are all classed as cloud. I think unless someone can definitively say this is incorrect I will leave it as its.

benjmarshall committed:
Update homeassistant/components/gotify/manifest.json

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

benjmarshall commented on homeassistant/components/gotify/manifest.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "domain": "gotify",
+  "name": "Gotify",
+  "documentation": "https://www.home-assistant.io/integrations/notify_events",
```
Thanks, good spot.

benjmarshall committed:
Update old function name

benjmarshall committed:
Use external library for service communication

benjmarshall committed:
Migrate component to config flow.

milanmeu committed:
Simplify ConfigFlow

benjmarshall committed:
Merge pull request #5 from milanmeu/patch-9

Simplify the ConfigFlow

benjmarshall committed:
Update tests to be compatible with simplified config flow

benjmarshall committed:
Merge pull request #6 from benjmarshall/gotify_config_flow

Migrate component to Config Flow

milanmeu commented on homeassistant/components/gotify/strings.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "title": "gotify",
```
The name from the manifest is automatically used as title, only set the title in strings.json if it should be translated or if it is different from the manifest.
```suggestion
```

benjmarshall commented:
@milanmeu you mentioned previously that I should change the way that the platform setup is done. Could you open a PR or provide some guidance on how I should be doing it? Thanks.

milanmeu commented on homeassistant/components/gotify/strings.json:
```diff
@@ -0,0 +1,22 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
Can we set a config flow unique id?
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids
If this is not possible then delete the unused already_configured string.

benjmarshall replied:
I don't think a unique id is suitable in this case having read through the linked documentation. I will remove this string.

benjmarshall committed:
Remove unused strings.

milanmeu commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,23 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up gotify_temp from a config entry."""
+    hass.data[DOMAIN] = entry.as_dict()
```
This will be overwritten if there are multiple config entries.
Use `hass.data.setdefault(DOMAIN, {})[entry.entry_id] = ...`.

benjmarshall replied:
How do i get access to the entry_id again inside the get_service function? I think I tried to use this method before but realised that a blank config was being passed to my get_service function since I moved to using Config Flow.

benjmarshall replied:
@milanmeu is this possible as I don't think the Notify platform has an async_setup_entry method as per documentation here: https://developers.home-assistant.io/docs/config_entries_index/#for-platforms

milanmeu replied:
> How do i get access to the entry_id again inside the get_service function? I think I tried to use this method before but realised that a blank config was being passed to my get_service function since I moved to using Config Flow.

Correct

> @milanmeu is this possible as I don't think the Notify platform has an async_setup_entry method as per documentation here: https://developers.home-assistant.io/docs/config_entries_index/#for-platforms

Correct

---

I'm sorry, I can't help you with this because my knowledge of services is not sufficient.
Can we have a service for a domain setup multiple times?

milanmeu commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,23 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return True
```
The GotifyNotificationService and `hass.data[DOMAIN][entry.entry_id]` should be removed.
For the service use: `hass.services.async_remove(...)`.

milanmeu commented on homeassistant/components/gotify/translations/en.json:
```diff
@@ -0,0 +1,22 @@
+{
```
```bash
python3 -m script.translations develop --integration gotify
```

benjmarshall replied:
Whoops, forgot that bit.

benjmarshall committed:
Fix and update translation strings

benjmarshall committed:
Fix calling discovery for Notification platform. Fix creation of specific notify service. Add intiial support for multiple entries.

benjmarshall commented:
I've updated this so that it should now correctly load the Notification platform and expose a "notify.your_gotify_entry" service.
I would appreciate a further code review at this point to make sure I've done this in the best way possible given the limitations of the Notify component.

The major outstanding issue with this integration is that multiple entries will not work correctly. The is due to the underlying library only exposing a bunch of functions, and using global variables for host/application ID etc, rather than defining a class which can hold a proper configuration for a specific host/application id. I might be able to hack around this by always re-configuring the host/application id with each message to be sent but this is messy. As a better alternative I have [opened a PR on the underlying library](https://github.com/d-k-bo/python-gotify/pull/1) which converts it to using a class. That way each specific GotifyNotificationService would be able to properly hold it's own configuration.

milanmeu reviewed (commented):
> The major outstanding issue with this integration is that multiple entries will not work correctly.

If we would accept this, you should add the following to the beginning of async_step_user:

```python
if self._async_current_entries():
    return self.async_abort(reason="single_instance_allowed")
```

benjmarshall committed:
Update interaction with gotify lib to use new class support

benjmarshall committed:
Fix default form input.

benjmarshall committed:
Update tests to match updated config flow

benjmarshall committed:
Change class to cloud_polling

benjmarshall committed:
Add unload function to remove notify service when config entry is removed.

benjmarshall commented:
The underlying library has now been updated with my suggested changes to implement a class. I've updated the integration here to use this updated library version so the integration now works correctly with multiple config entries. And I've managed to properly unload the notify service when an entry is removed, thanks @milanmeu.

I'd appreciate someone providing what I hope to be a final review/test before we can call this integration good to go. I'll be updating the doc's next to reflect the new config flow steps.

benjmarshall commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,98 @@
+            {
+                vol.Required(CONF_HOST, default=user_input.get(CONF_HOST)): str,
+                vol.Required(CONF_TOKEN, default=user_input.get(CONF_TOKEN)): str,
+                vol.Required(CONF_NAME, default=user_input.get(CONF_NAME)): str,
```
Thanks, updated.

milanmeu commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,34 @@
+from .const import DOMAIN
+
+
+async def async_setup(hass, config):
+    """Set up the gotify component."""
+    hass.data[DOMAIN] = config
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up gotify from a config entry."""
+    hass.data[DOMAIN][entry.data[CONF_NAME]] = entry
```
```suggestion
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up gotify from a config entry."""
    hass.data.setdefault(DOMAIN, {})[entry.data[CONF_NAME]] = entry
```

milanmeu commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,34 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    hass.services.async_remove("notify", entry.data["name"])
```
```suggestion
    hass.services.async_remove("notify", entry.data[CONF_NAME])
    hass.data[DOMAIN].pop(entry.data[CONF_NAME])
```

benjmarshall committed:
Update setup and unload functions with suggested changes from review

benjmarshall commented:
Thanks @milanmeu, updated. I've noticed just now that when using the integration page option to "rename" an integration entry the name of the notify service does not get updated. Do you think i should try and hook into this rename call somehow to recreate the notify service? Is there a method I can add to the config flow to do this?

benjmarshall commented:
Docs updated.

milanmeu commented:
> I've noticed just now that when using the integration page option to "rename" an integration entry the name of the notify service does not get updated. 

The title of the config entry is independent of the service name. You don't have to change this.

timdonovanuk commented:
Thanks for this @benjmarshall !

TheLastProject commented:
Given the bot marked this stale, is there any way anyone can help move this forward? I started using Gotify recently and would be interested in a more native integration. Are there any issues remaining keeping this from being merged?

benjmarshall commented:
As far as I'm concerned this is ready to be merged. It's functional, it's been reviewed and I've been using it locally for several months now without issues. 

@milanmeu @TheZoker Do you know how I can help get this merged? I'm happy to merge in the latest changes on the dev branch if that helps? Is there someone I can contact who can sign this off for being merged?

Joshndroid commented:
Bumping this agfain as its been marked at stale... as I would also like gotify integration thanks heaps @benjmarshall

flabbamann commented:
I would like to see this getting integrated too. Could someone activate the CI pipeline here so we see if there are todos left? 😃

benjmarshall commented:
> I would like to see this getting integrated too. Could someone activate the CI pipeline here so we see if there are todos left? smiley

Is that something I can do?

Joshndroid commented:
@milanmeu @TheZoker Can we get this merged?
Using android devices without play services/google, you cant use FCM for messages (well it doesnt seem to like it for me anyway).
This is the perfect work around for me to receive notifications.

bdraco commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,98 @@
+        STEP_USER_DATA_SCHEMA = vol.Schema(
+            {
+                vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
+                vol.Required(CONF_TOKEN, default=user_input.get(CONF_TOKEN, "")): str,
```
We shouldn't send the token back to the user once they have entered it if they end up having to change something.

bdraco commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,98 @@
+            {
+                vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
+                vol.Required(CONF_TOKEN, default=user_input.get(CONF_TOKEN, "")): str,
+                vol.Required(CONF_NAME, default=user_input.get(CONF_NAME, "")): str,
```
Usually we try to avoid collecting a name unless it is needed to setup the integration as we try to get that from the device or service.

bdraco commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,98 @@
+        else:
+            user_input = {}
+
+        STEP_USER_DATA_SCHEMA = vol.Schema(
```
Usually we only use UPPER for constants.  This should be renamed or an anonymous inline below

bdraco commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,28 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up gotify from a config entry."""
+    hass.data.setdefault(DOMAIN, {})[entry.data[CONF_NAME]] = entry
```
We usually store integration data as `hass.data.setdefault(DOMAIN, {})[entry.entry_id]`

benjmarshall committed:
Use entry_id as ID to store config against. Re-work initial configuration to use client token to allow the creationg of a new gotify application if required. Re-work to allow multiple config entires to exist. Generate entry name from gotify application name instead of user supplied name.

benjmarshall commented:
Hi All, I've re-worked this component a little today on the basis of comments from @bdraco.

First up the initial configuration has been re-worked to use a Gotify client token, instead of an app token. This allows access to more API endpoints which allows the flexibility to get application details and create new applications if required.

An application token can still be optionally supplied in the config. If supplied then a config entry will be created for the Gotify application that corresponds to this token. If no application token is supplied then a new Gotify application is created on the Gotify server.

The name of the config entry is now generated based on the name of the Gotify application (either created or determined by the supplied app token). Because the name of the config entry is used to create a notify service this needs to be unique. I have added a check to make sure that a unique config entry name is used. A postscript integer (starting with 2) will be appended if the Gotify application would not create a unique config entry name. 

These changes should address all of the comments from @bdraco. Additionally these changes allow for multiple config entries to co-exist gracefully.

Due to the fact I have changed more than strictly the suggested changes by @bdraco it would be beneficial to have a further review by the previous comments @milanmeu @TheZoker.

These changes will also require an update to the documentation which I will push soon.

benjmarshall committed:
Remove print statement.

TheLastProject commented:
Bump for bot (because AFAICT this is again waiting on the Home Assistant team)

jackwilsdon commented:
If it helps at all, I've been using this as a custom component for a few months now and it has been rock solid.

jackwilsdon commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+    raw_name = DOMAIN + "_" + name.replace(" ", "_").lower()
+
+    if DOMAIN in hass.data:
+        entry_names = []
```
I think this needs moving outside of this `if` statement (or the `while` loop needs moving inside of it).

ViViDboarder replied:
Yea, this is a bug for sure. Since the lower half of the function doesn't do anything really if `entry_names` is empty, I'd recommend a short circuit exit.

```python
if DOMAIN not in hass.data:
    return name
```

jackwilsdon commented:
@Hukuma1 sure:
1. Download Ben's branch: https://github.com/benjmarshall/core/archive/refs/heads/add_gotify_integration.zip
2. Copy `homeassistant/components/gotify` into your `custom_components` directory (so you end up with `custom_components/gotify` with all of the contents of `homeassistant/components/gotify`)
3. Add a `version` key to the component's `manifest.json`:
   ```diff
   --- a/manifest.json
   +++ b/manifest.json
   @@ -1,6 +1,7 @@
    {
      "domain": "gotify",
      "name": "Gotify",
   +  "version": "1.0.0",
      "documentation": "https://www.home-assistant.io/integrations/gotify",
      "codeowners": ["@benjmarshall"],
      "iot_class": "cloud_polling",
   ```
4. Make this change to `config_flow.py` (I've also left a review comment about this):
   ```diff
   --- a/config_flow.py
   +++ b/config_flow.py
   @@ -46,8 +46,8 @@
    
        raw_name = DOMAIN + "_" + name.replace(" ", "_").lower()
    
   +    entry_names = []
        if DOMAIN in hass.data:
   -        entry_names = []
            for entry in hass.data[DOMAIN]:
                entry_names.append(hass.data[DOMAIN][entry].data[CONF_NAME])
   ```

If you have any issues, feel free to message me on the Home Assistant Community (either using [this direct link](https://community.home-assistant.io/new-message?username=jackwilsdon) or via [my profile](https://community.home-assistant.io/u/jackwilsdon)) as I don't wish to start any unrelated discussions on this PR.

ViViDboarder commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    discovery_info = {CONF_NAME: entry.data[CONF_NAME], "entry_id": entry.entry_id}
+    await hass.async_create_task(
+        discovery.async_load_platform(
+            hass, "notify", DOMAIN, discovery_info, hass.data[DOMAIN]
```
The string, `notify`, should be imported from `homeassistant.const.Platform`.

Eg.
```python
from homeassistant.const import Platform

Platform.NOTIFY
```

ViViDboarder commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    # Gotify notify component requires entry_id in discovery_info to get the correct entry config.
+    # The discovery platform requires CONF_NAME in discovery_info to set a service name.
+    discovery_info = {CONF_NAME: entry.data[CONF_NAME], "entry_id": entry.entry_id}
+    await hass.async_create_task(
```
I think the typical pattern for this has changed to:

```python
PLATFORMS = [Platform.NOTIFY]  # Somewhere above in the module

hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

I'm not sure if that warrants changing or not.
```

milanmeu replied:
The notify platform doesn't support config entries yet. You have to use discovery to load platform.

ViViDboarder replied:
Ah, good to know!

ViViDboarder commented on homeassistant/components/gotify/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    hass.services.async_remove("notify", entry.data[CONF_NAME])
+    hass.data[DOMAIN].pop(entry.entry_id)
+    return True
```
Now the standard pattern here is

```python
async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
```

Again, what you have probably works, but I'm not sure if updating to the new methods is required for merging.

milanmeu replied:
Same as above the notify platform doesn't support config entries.

ViViDboarder replied:
Yea, makes sense.

ViViDboarder commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+
+    if DOMAIN in hass.data:
+        entry_names = []
+        for entry in hass.data[DOMAIN]:
+            entry_names.append(hass.data[DOMAIN][entry].data[CONF_NAME])
```
As this is a dictionary and you only seem to care about the values, you can instead iterate over `hass.data[DOMAIN].values()` directly.

You can also simplify this using list comprehension and even avoid empty list initialization.

```python
entry_names = [entry.data[CONF_NAME] for entry in hass.data[DOMAIN].values()]
```

ViViDboarder commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+        for entry in hass.data[DOMAIN]:
+            entry_names.append(hass.data[DOMAIN][entry].data[CONF_NAME])
+
+    name = raw_name
+    i = 2
+    while name in entry_names:
+        name = raw_name + str(i)
+    return name
```
This will cause an infinite loop if the same name exists since you never remove the name from the `entry_names` list.

You're probably better off filtering by name in the list above and then using it's length here.

```
entry_names = [entry.data[CONF_NAME] for entry in hass.data[DOMAIN].values() if entry.data[CONF_NAME] == name]
name = raw_name + str(len(entry_names))
```

That said, we're not using the actual name values at all, so it could be simplified further.

```python
def sanitise_name(hass: HomeAssistant, name: str) -> str:
    """Make sure we are not going to create and entry with a duplicate name as this would produce duplicate services."""
    num_existing_names = 1
    for entry in hass.data.get(DOMAIN, {}).values():
        if entry.data[CONF_NAME] == name:
            num_existing_names += 1

    name = DOMAIN + "_" + name.replace(" ", "_").lower()
    if num_existing_names == 1:
        return name

    return f"{name}_{num_existing_names}"
```

ViViDboarder commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+    return_name = ""
+
+    try:
+        current_applications = await hass.async_add_executor_job(
```
It looks like this is only used in one block below. It can go within the `if CONF_TOKEN in user_input:` block.

ViViDboarder commented on homeassistant/components/gotify/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+                if user_input[CONF_TOKEN] == app.get("token"):
+                    return_token = user_input[CONF_TOKEN]
+                    return_name = app.get("name")
+        if not (CONF_TOKEN in user_input) or not return_token:
```
You could drop `not (CONF_TOKEN in user_input)` and only check `not return_token` here as `return_token` is initialized as a falsy value.

ViViDboarder commented on tests/components/gotify/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    ) as mock_setup_entry:
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {"host": "https://1.1.1.1", "token": "test-token", "name": "test"},
```
Minor nit: 1.1.1.1 is a real public IP address. [`example.com`](https://www.iana.org/domains/reserved) is reserved for these purposes.

ViViDboarder commented on tests/components/gotify/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {"host": "1.1.1.1", "token": "test-token", "name": "test"},
```
This is a semantic thing, but technically this is a host address, though it is not a URI. https://www.sistrix.com/ask-sistrix/technical-seo/site-structure/what-is-the-difference-between-a-url-domain-subdomain-hostname-etc

ViViDboarder reviewed (changes requested):
A few small bugs and then some recommendations.

Pimmetje commented:
Bump

fdw commented:
As anyone still working on this PR (especially @benjmarshall)?

benjmarshall commented:
I don't think anyone is actively working on this. Whilst I am using it on my install as a custom component I am not actively developing it as it provides the functionality I need already. I have pretty much given up trying to get it merged as no one from the core team with permissions to do so seemed interested. I am sympathetic as I imagine they would rather focus on their integrated mobile app notifications.

The component still works for me so if you have the knowledge it can be installed manually as a custom component. I should however say that my install is behind the latest Home Assistant by a bit now so no guarantees on compatibility with the latest release(s).

13 May 2023 08:31:31 Fabian Winter ***@***.***>:

> As anyone still working on this PR (especially @benjmarshall[https://github.com/benjmarshall])?
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/home-assistant/core/pull/53050#issuecomment-1546585951], or unsubscribe[https://github.com/notifications/unsubscribe-auth/ABQEWSFUEKUT2ON4XZGVVH3XF42E5ANCNFSM5ANAOVHA].
> You are receiving this because you were mentioned.[Tracking image][https://github.com/notifications/beacon/ABQEWSC2TZWCJ3QCGC36A2TXF42E5A5CNFSM5ANAOVHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOLQXQOXY.gif]
>

marcelveldt commented:
I've marked this PR as draft, as changes are requested that need to be processed first before we can do another review.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! :thumbsup: 

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
